torch/csrc/jit/import.cpp

```diff
@@ -155,27 +161,162 @@ void buildBlock(const onnx_torch::GraphProto& graph_proto, Block* block,
+  auto model_proto = onnx_torch::ModelProto();
+  model_proto.ParseFromString(serialized_graph);
 
-  for (auto tensor_ : graph_proto.initializer()) {
+  auto graph_proto = model_proto.graph();
+  auto graph = buildGraph(graph_proto);
+  for (auto &tensor_ : graph_proto.initializer()) {
     initializers.push_back(buildTensor(tensor_));
   }
-
   return graph;
 }
 
+class ModuleDecoder : JitDecoder {
+ public:
+  std::shared_ptr<script::Module> decode(
+      std::shared_ptr<script::Module> root_module,
+      const std::string& serialized_module,
+      const std::unordered_map<std::string, std::string>& storage_map);
+
+ private:
+  at::Tensor buildTensor(const onnx_torch::TensorProto& tensor_proto);
+
+  at::Tensor buildParameter(const onnx_torch::TensorProto& tensor_proto);
+
+  at::Storage* getStorage(at::ScalarType type, std::string name);
+
+  std::pair<std::shared_ptr<script::Module>, std::string> parseFullName(
+      std::shared_ptr<script::Module> root_module,
+      const std::string fullname);
+
+  const std::unordered_map<std::string, std::string> *storage_export_map_;
+  std::unordered_map<std::string, std::unique_ptr<at::Storage>> storage_map_;
+};
+
+at::Tensor ModuleDecoder::buildParameter(const onnx_torch::TensorProto& tensor_proto) {
+  auto storage = getStorage(onnxTypeToATenType(tensor_proto.data_type()), tensor_proto.doc_string());
+  std::vector<int64_t> dims, strides;
+  std::move(tensor_proto.dims().begin(), tensor_proto.dims().end(), std::back_inserter(dims));
+  std::move(tensor_proto.int32_data().begin() + 3, tensor_proto.int32_data().end(), std::back_inserter(strides));
+  auto tensor = at::CPU(onnxTypeToATenType(tensor_proto.data_type())).tensor(
+      *storage, tensor_proto.int64_data(2), dims, strides);
+  autograd::Variable var = autograd::make_variable(tensor, tensor_proto.int64_data(0));
+  return var;
 }
 
-std::shared_ptr<Graph> ImportIRGraph(const std::string& serialized_graph,
-                                     std::vector<at::Tensor>& initializers) {
+at::Tensor ModuleDecoder::buildTensor(const onnx_torch::TensorProto& tensor_proto) {
+  auto storage = getStorage(onnxTypeToATenType(tensor_proto.data_type()), tensor_proto.doc_string());
+  std::vector<int64_t> dims, strides;
+  std::move(tensor_proto.dims().begin(), tensor_proto.dims().end(), std::back_inserter(dims));
+  std::move(tensor_proto.int32_data().begin() + 1, tensor_proto.int32_data().end(), std::back_inserter(strides));
+  auto tensor = at::CPU(onnxTypeToATenType(tensor_proto.data_type())).tensor(
+      *storage, tensor_proto.int64_data(0), dims, strides);
+  return tensor;
+}
+
+at::Storage* ModuleDecoder::getStorage(at::ScalarType type, std::string name) {
+  auto storage_it = storage_map_.find(name);
+  if (storage_it == storage_map_.end()) {
+    auto storage = at::CPU(type).storage();
+    auto string_it = storage_export_map_->find(name);
+    JIT_ASSERT(string_it != storage_export_map_->end());
```

Comment:
Rather than assert here, can't we just initialize this entry the first time it is requested? That we there is no "action at a distance"

Reply:
storage_export_map_ contains all the things that were loaded in. The assert is just a sanity check to make sure that the storage we're looking for is in the set of storages we loaded in.